### PR TITLE
[JENKINS-39835] - Be extra defensive about Errors and Exceptions

### PR DESCRIPTION
**Pull request by jtnord**
[JENKINS-39835](https://issues.jenkins-ci.org/browse/JENKINS-39835) Be even more defensive then against leaving connections dangling.

NB: no testing performed of this fix - it is a direct port from the internal code that was submitted to make the JNLP v4 protocol.
@reviewbybees esp @stephenc @oleg-nenashev

### CODE REVIEW

**Comment by stephenc** [comment]
:bee:

File: src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java [hunk 1]
```diff
@@ -177,15 +177,19 @@ public void ready(boolean accept, boolean connect, boolean read, boolean write)
...
+                        try {
+                            if (LOGGER.isLoggable(Level.SEVERE)) {
+                                LogRecord record = new LogRecord(Level.SEVERE, "[{0}] Uncaught {1}");
+                                record.setThrown(t);
+                                record.setParameters(new Object[]{stack().name(), t.getClass().getSimpleName()});
+                            }
+                        } finally {
+                            // incase this was an OOMErr and logging caused another OOMErr
+                            recvKey.cancel();
+                            onRecvClosed();
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java]: I suppose this code should rethrow errors afterwards

Comment (amuniz) [inline src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java]: It was not doing it before the change.

**Comment by amuniz** [comment]
:bee:

File: src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java]: It was not catching `Error`s before the change

**Comment by oleg-nenashev** [comment]
@jtnord Do you think we need a similar fix in stable-2.x?

File: src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java (hunk 1 above)
Comment (jtnord) [inline src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java]: if it does that you could kill the thread / leave the system in a bad state.

**Comment by jtnord** [comment]
@oleg-nenashev if you throw the error will likely kill the thread and you may not end up in the state you think you want to end up in.

I'm not comfortable re-throwing Errors, so if you would like that then I will leave it up to yourself, cloudbees oss-team or someone else to provide.

**Comment by jtnord** [comment]
> Do you think we need a similar fix in stable-2.x?

This does not contain the JNLP4v protocol (and hence this code) does it?

**Comment by stephenc** [comment]
W.r.t. Killing the thread. Thinking about this more... You would only ever kill a worker thread, which should be ok as they are just assigned to this operation and the threadpool will replace if they die

**Comment by oleg-nenashev** [comment]
@jtnord @stephenc Should we proceed with this PR?

**Comment by jtnord** [comment]
Yes.. ops read is no longer registered and only adds it later on, leaving hanging network.  Seen at a customer site.

**Comment by oleg-nenashev** [comment]
I still do not like the handling of some fatal errors, but I agree to merge it taking the potential impact into account.

**Comment by oleg-nenashev** [comment]
BTW I also think that killing the worker thread is fine

**Comment by jtnord** [comment]
@oleg-nenashev if the worker thread dies handling a read then whatever is connected is dead as far as OP_READ.  as OP_READ is removed when we start the thread and only removed when we have read all of the data available in the callback (ie the thread that dies).

See line 147 in [this block](https://github.com/jenkinsci/remoting/blob/ec9b5c13b879f44c04fa28ee6c8b113a165c9e57/src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java#L138-L165)

Did you see my proposal on the Jenkins-dev list for handling uncaught exceptions :-)

**Comment by oleg-nenashev** [comment]
@jtnord 
> Did you see my proposal on the Jenkins-dev list for handling uncaught exceptions 

Show me the link

File: src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java [hunk 2]
```diff
@@ -177,15 +177,19 @@ public void ready(boolean accept, boolean connect, boolean read, boolean write)
...
-                        if (LOGGER.isLoggable(Level.WARNING)) {
-                            LogRecord record = new LogRecord(Level.WARNING, "[{0}] Uncaught {1}");
-                            record.setThrown(e);
-                            record.setParameters(new Object[]{stack().name(), e.getClass().getSimpleName()});
+                        try {
+                            if (LOGGER.isLoggable(Level.SEVERE)) {
+                                LogRecord record = new LogRecord(Level.SEVERE, "[{0}] Uncaught {1}");
+                                record.setThrown(t);
+                                record.setParameters(new Object[]{stack().name(), t.getClass().getSimpleName()});
+                            }
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java]: The whole block could be written more simply as

```java
LOGGER.log(Level.SEVERE, "Uncaught error in " + stack().name(), t);
```

without any real loss.

File: src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java [hunk 3]
```diff
@@ -177,15 +177,19 @@ public void ready(boolean accept, boolean connect, boolean read, boolean write)
...
+                        } finally {
+                            // incase this was an OOMErr and logging caused another OOMErr
```
Comment (jglick) [inline src/main/java/org/jenkinsci/remoting/protocol/impl/NIONetworkLayer.java]: Easier and probably safer to set a `boolean` flag at the top of the method indicating whether it completed normally; wrap everything in a `try`-`finally` block that closes the channel if not.

**Comment by oleg-nenashev** [comment]
OK, let's go forward with this PR as is.
@jtnord , my plan is to have remoting-3.2.1 with this fix only if @olivergondza agrees. Actually #138 may also deserve backporting due to the quite wide exposure of the issue. @olivergondza WDYT?

**Comment by olivergondza** [comment]
I am ok with the code as it is - logging the errors is fine. It seems reasonable safe to have it backported into .2.

**Comment by oleg-nenashev** [comment]
OK, going forward with the merge. Will be in this weekly if @kohsuke is available

**Pull request merged**